apps/src/javalab/Javalab.js

```diff
@@ -388,7 +388,13 @@ Javalab.prototype.executeJavabuilder = function(executionType) {
   );
 
   let connectToJavabuilder;
-  if (this.isEditingExemplar || this.isViewingExemplar) {
+  // If in exemplar or preview mode, use the sources currently on the page
+  // (as opposed to the sources saved on the backend).
+  if (
+    this.isEditingExemplar ||
+    this.isViewingExemplar ||
+    getStore().getState().pageConstants.isReadOnlyWorkspace
```

Comment:
out of scope for this pr, but I am wondering if we should ever use the sources saved on the backend? We probably always want to run what is showing to the user, which is often what is also saved on the backend. But if they aren't the same we should go with what the user sees.

Reply:
Yeah good question - IIRC we considered it while designing the decoupled javabuilder solution, but ultimately decided not to, due to us wanting to hide validation code from the user. But looks like we supply validation from the client now as well, so not sure if that's still a concern...

Reply:
oh right that's why. I think sources could always come from the client but it's not a high priority

Reply:
Yeah agreed - I'll add a task

Reply:
If I'm remembering right, we decided to use sources from the backend for a few reasons:
* validation code, as mentioned
* too large of a payload - student projects in App Lab regularly grow to low thousands of lines of code, we suspected the same would happen in Java Lab. With the non-decoupled version of Javabuilder, we would have quickly surpassed payload limits.
* consistency between the backend and the code that is run. This is an issue we run into quite often in certain labs, such as Web Lab. Sometimes an issue happens with the PUT so the code that is executed is different from what's saved on S3. Then, when the student reloads the page, they have a different experience from the last time they were viewing their code.
^ I think so long as we wait for the save to complete successfully before running the code, that last item shouldn't be an issue.

Meta point: this is moreso me thinking out loud and providing context. I was initially a bit concerned about this change, but couldn't put my finger on why. Writing these out made me realize it was bullet 2 (already solved for) and bullet 3 (unlikely to be an issue for us). So this ends up being a less useful comment, but I'll go ahead and leave it to provide context in the future if needed.